core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/SingleDirectionalParentVariableReferenceGraph.java

```diff
@@ -0,0 +1,134 @@
+        this.changedVariableNotifier = changedVariableNotifier;
+        var shadowEntities = Arrays.stream(entities).filter(entityClass::isInstance).toArray();
+        var loopedDescriptor =
+                sortedDeclarativeShadowVariableDescriptors.get(0).getEntityDescriptor().getShadowVariableLoopedDescriptor();
+
+        var updaterIndex = 0;
+        for (var variableDescriptor : sortedDeclarativeShadowVariableDescriptors) {
+            var variableMetaModel = variableDescriptor.getVariableMetaModel();
+            var variableUpdaterInfo = new VariableUpdaterInfo<>(
+                    variableMetaModel,
+                    variableDescriptor,
+                    loopedDescriptor,
+                    variableDescriptor.getMemberAccessor(),
+                    variableDescriptor.getCalculator()::executeGetter);
+            sortedVariableUpdaterInfos[updaterIndex++] = variableUpdaterInfo;
+
+            for (var source : variableDescriptor.getSources()) {
+                for (var sourceReference : source.variableSourceReferences()) {
+                    monitoredSourceVariableSet.add(sourceReference.variableMetaModel());
+                }
+            }
+        }
+
+        for (var shadowEntity : shadowEntities) {
+            updateChanged(shadowEntity);
+        }
+        if (loopedDescriptor != null) {
+            for (var shadowEntity : shadowEntities) {
+                changedVariableNotifier.beforeVariableChanged().accept(loopedDescriptor, shadowEntity);
+                loopedDescriptor.setValue(shadowEntity, false);
+                changedVariableNotifier.afterVariableChanged().accept(loopedDescriptor, shadowEntity);
+            }
+        }
+    }
+
+    @Override
+    public @Nullable EntityVariablePair<Solution_> lookupOrNull(VariableMetaModel<?, ?, ?> variableId, Object entity) {
+        throw new IllegalStateException("Impossible state: cannot lookup in a %s graph."
+                .formatted(SingleDirectionalParentVariableReferenceGraph.class.getSimpleName()));
+    }
+
+    @Override
+    public void addEdge(@NonNull EntityVariablePair<Solution_> from, @NonNull EntityVariablePair<Solution_> to) {
+        throw new IllegalStateException("Impossible state: cannot modify an %s graph."
+                .formatted(SingleDirectionalParentVariableReferenceGraph.class.getSimpleName()));
+    }
+
+    @Override
+    public void removeEdge(@NonNull EntityVariablePair<Solution_> from, @NonNull EntityVariablePair<Solution_> to) {
+        throw new IllegalStateException("Impossible state: cannot modify an %s graph."
+                .formatted(SingleDirectionalParentVariableReferenceGraph.class.getSimpleName()));
+    }
+
+    @Override
+    public void markChanged(@NonNull EntityVariablePair<Solution_> node) {
+        throw new IllegalStateException("Impossible state: cannot mark changed an %s graph."
+                .formatted(SingleDirectionalParentVariableReferenceGraph.class.getSimpleName()));
+    }
```

Comment:
Since the "graph" is already part of the class name, I'd simplify the message to remove that word.